### ISSUE TITLE
Crash in osg::State:applyAttribute(...) 

### ISSUE THREAD

**Issue by remoe** [gwaldron/osgearth#902]
Platform: 

MacOSX Core Profile (GL3)
osgEarth: 208ab3d922c6fbca5844ace4379f62d7d51b7255 and 8bef1235f84d9242324d12e09326d5952c4f4552
OSG: 3.5.6 master

Sometimes i have following crashes:

```#0 0x0000000101719e92 in osg::State::applyAttribute(osg::StateAttribute const*, osg::State::AttributeStack&) at /Users/user/src/sdk/osg/3.5.6/build/osx/include/osg/State:1146
#1  0x000000010170c15b in osg::State::applyAttributeList(std::__1::map<std::__1::pair<osg::StateAttribute::Type, unsigned int>, osg::State::AttributeStack, std::__1::less<std::__1::pair<osg::StateAttribute::Type, unsigned int> >, std::__1::allocator<std::__1::pair<std::__1::pair<osg::StateAttribute::Type, unsigned int> const, osg::State::AttributeStack> > >&, std::__1::map<std::__1::pair<osg::StateAttribute::Type, unsigned int>, std::__1::pair<osg::ref_ptr<osg::StateAttribute>, unsigned int>, std::__1::less<std::__1::pair<osg::StateAttribute::Type, unsigned int> >, std::__1::allocator<std::__1::pair<std::__1::pair<osg::StateAttribute::Type, unsigned int> const, std::__1::pair<osg::ref_ptr<osg::StateAttribute>, unsigned int> > > > const&) at /Users/user/src/sdk/osg/3.5.6/build/osx/include/osg/State:1897
#2  0x00000001016f16c0 in osg::State::apply(osg::StateSet const*) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osg/State.cpp:666
#3  0x0000000100bbd495 in osgUtil::RenderLeaf::render(osg::RenderInfo&, osgUtil::RenderLeaf*) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderLeaf.cpp:47
#4  0x0000000100ba8ca9 in osgUtil::RenderBin::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderBin.cpp:467
#5  0x0000000100ba856a in osgUtil::RenderBin::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderBin.cpp:430
#6  0x0000000100ba9897 in osgUtil::RenderBin::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderBin.cpp:517
#7  0x0000000100bc9cfb in osgUtil::RenderStage::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderStage.cpp:1401
#8  0x0000000100ba856a in osgUtil::RenderBin::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderBin.cpp:430
#9  0x0000000100bc6e6d in osgUtil::RenderStage::drawInner(osg::RenderInfo&, osgUtil::RenderLeaf*&, bool&) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderStage.cpp:931
#10 0x0000000100bc9472 in osgUtil::RenderStage::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/RenderStage.cpp:1242
#11 0x0000000100be68e3 in osgUtil::SceneView::draw() at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgUtil/SceneView.cpp:1367
#12 0x00000001021c73e9 in osgViewer::Renderer::draw() at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgViewer/Renderer.cpp:773
#13 0x00000001021c8e71 in osgViewer::Renderer::operator()(osg::GraphicsContext*) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osgViewer/Renderer.cpp:933
#14 0x00000001015afae5 in osg::GraphicsContext::runOperations() at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osg/GraphicsContext.cpp:695
#15 0x00000001015c2ae8 in osg::RunOperations::operator()(osg::GraphicsContext*) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osg/GraphicsThread.cpp:139
#16 0x00000001015c21d0 in osg::GraphicsOperation::operator()(osg::Object*) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osg/GraphicsThread.cpp:53
#17 0x00000001016587ba in osg::OperationThread::run() at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osg/OperationThread.cpp:438
#18 0x00000001015c20dc in osg::GraphicsThread::run() at /Users/user/src/sdk/osg/3.5.6/build/osx/src/osg/GraphicsThread.cpp:38
#19 0x00000001039971a4 in OpenThreads::ThreadPrivateActions::StartThread(void*) at /Users/user/src/sdk/osg/3.5.6/build/osx/src/OpenThreads/pthreads/PThread.cpp:214
``` 

```c++
            if (as.last_applied_attribute != attribute)
            {
                if (!as.global_default_attribute.valid()) as.global_default_attribute = attribute->cloneType()->asStateAttribute();

                as.last_applied_attribute = attribute;
>>>                attribute->apply(*this);

                const ShaderComponent* sc = attribute->getShaderComponent();
                if (as.last_applied_shadercomponent != sc)
                {
                    as.last_applied_shadercomponent = sc;
                    _shaderCompositionDirty = true;
                }

                if (_checkGLErrors==ONCE_PER_ATTRIBUTE) checkGLErrors(attribute);

                return true;
            }
```

```
(lldb) print *attribute
(const osg::StateAttribute) $2 = {
  osg::Object = {
    osg::Referenced = {
      _observerSet = (_ptr = 0x000060000010e7c0)
      _refCount = (_value = 1431655765)
    }
    _name = ""...
    _dataVariance = 1431655765
    _userDataContainer = 0x5555555555555555
  }
  _parents = size=0 {}
  _shaderComponent = {
    _ptr = 0x5555555555555555
  }
  _updateCallback = {
    _ptr = 0x5555555555555555
  }
  _eventCallback = {
    _ptr = 0x5555555555555555
  }
}

```

```
print _attributeMap
(osg::State::AttributeMap) $0 = size=10 {
  [0] = {
    first = (first = POLYGONMODE, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x00006000002e9280
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x00006000002e9280
      }
      attributeVec = size=0 {}
    }
  }
  [1] = {
    first = (first = MATERIAL, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x000000010bc78790
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x000000010ba11590
      }
      attributeVec = size=1 {
        [0] = {
          first = 0x000000010bc78790
          second = 0
        }
      }
    }
  }
  [2] = {
    first = (first = CULLFACE, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x00006000002e8d80
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x00006000002e8d80
      }
      attributeVec = size=0 {}
    }
  }
  [3] = {
    first = (first = LIGHT, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x00006000001a3720
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x00006000001a3720
      }
      attributeVec = size=0 {}
    }
  }
  [4] = {
    first = (first = LINEWIDTH, second = 0)
    second = {
      changed = true
      last_applied_attribute = 0x00006080004ebb80
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x00006080002ebc80
      }
      attributeVec = size=0 {}
    }
  }
  [5] = {
    first = (first = BLENDFUNC, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x00006100000e4e80
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x00006080000e2d00
      }
      attributeVec = size=1 {
        [0] = {
          first = 0x00006100000e4e80
          second = 0
        }
      }
    }
  }
  [6] = {
    first = (first = COLORMASK, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x00006180000ebe80
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x00006080000e2d80
      }
      attributeVec = size=1 {
        [0] = {
          first = 0x00006180000ebe80
          second = 0
        }
      }
    }
  }
  [7] = {
    first = (first = DEPTH, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x0000600000301f80
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x000061800010e4f0
      }
      attributeVec = size=3 {
        [0] = {
          first = 0x0000600000302c70
          second = 0
        }
        [1] = {
          first = 0x0000600000507e90
          second = 0
        }
        [2] = {
          first = 0x0000600000301f80
          second = 0
        }
      }
    }
  }
  [8] = {
    first = (first = VIEWPORT, second = 0)
    second = {
      changed = false
      last_applied_attribute = 0x0000610000105100
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x0000608000104c80
      }
      attributeVec = size=2 {
        [0] = {
          first = 0x0000610000105100
          second = 0
        }
        [1] = {
          first = 0x0000610000105100
          second = 0
        }
      }
    }
  }
  [9] = {
    first = (first = PROGRAM, second = 0)
    second = {
      changed = true
      last_applied_attribute = 0x0000000000000000
      last_applied_shadercomponent = 0x0000000000000000
      global_default_attribute = {
        _ptr = 0x000000010a71b7d0
      }
      attributeVec = size=6 {
        [0] = {
          first = 0x000000010a752510
          second = 0
        }
        [1] = {
          first = 0x000000010bf48650
          second = 0
        }
        [2] = {
          first = 0x000000012b224af0
          second = 0
        }
        [3] = {
          first = 0x000000012b224230
          second = 0
        }
        [4] = {
          first = 0x000000012b223f30
          second = 0
        }
        [5] = {
          first = 0x000000012b2233b0
          second = 0
        }
      }
    }
  }
}
```

**Comment by remoe** [comment]
It doesn't crash when i disable DepthOffsetAdapter !

```c++
void
DepthOffsetAdapter::init()
{
    _supported = false; //Registry::capabilities().supportsGLSL();
    if ( _supported )
```

**Comment by remoe** [comment]
This changes "creates" the crash:

https://github.com/gwaldron/osgearth/commit/c8ffd05ee5dcd29ce2ab9105dc00144e36c15580

**Comment by jasonbeverage** [comment]
Hey Remo, I pushed a fix for this in 38f58fb1df6500376122fd41e50e76df4f08db63.  The DepthOffsetAdapter was installing itself on every tile load instead of just once.
